[Communication] - phone-numbers - Core V2 migration

danielortega-msft opened the pull request:
### Packages impacted by this PR
 - `@azure/communication-phone-numbers`

### Issues associated with this PR
- https://github.com/Azure/azure-sdk-for-js/issues/20056
- https://github.com/Azure/azure-sdk-for-js/issues/15814

### Describe the problem that is addressed by this PR
Migrate the HTTP pipeline to the new architecture in Azure Core V2. Refer to the following documentation for more details: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-rest-pipeline/documentation/core2.md

### What are the possible designs available to address the problem? If there are more than one possible design, why was the one in this PR chosen?
For information on the Core V2 migration, refer to: https://github.com/Azure/azure-sdk-for-js/blob/master/sdk/core/core-rest-pipeline/documentation/core2.md

This migration introduced the following two compatibility issues:
- The phone numbers API uses relative URLs for `nextLink` values in endpoints that support paging, which is not supported out-of-the-box in the new client. While the current REST API guidelines state that they should be absolute URLs, this was introduced after the initial release of the API (for more details, see the [PR discussion](https://github.com/microsoft/api-guidelines/pull/264#discussion_r665764432) that introduced the change).
- LROs in the phone numbers API have status-monitors given by the `operation-location` header; however, the `@azure/core-lro` implementation expects it to be in the legacy `azure-asyncoperation` header. While the current version of the guidelines discourage the later approach, they indicate that client code should look for both. For more details, refer to the [Azure REST API guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#long-running-operations-with-status-monitor).

In order to solve both issues, custom pipeline policies were implemented to make the HTTP responses have the form expected by the Core V2 libraries. These policies are workarounds and should be cleaned up once the incompatibilities are addressed.

### Are there test cases added in this PR? _(If not, why?)_
Yes, the test cases for the policies implemented as workarounds for compatibility issues.

### Provide a list of related PRs _(if any)_
- https://github.com/Azure/azure-sdk-for-js/pull/20337

### Checklists
- [x] Added impacted package name to the issue description
- [ ] Does this PR needs any fixes in the SDK Generator?** _(If so, create an Issue in the [Autorest/typescript](https://github.com/Azure/autorest.typescript) repository and link it here)_
- [x] Added a changelog (if necessary)

danielortega-msft commented:
FYI, the following issue was created to track the LRO incompatibility: https://github.com/Azure/azure-sdk-for-js/issues/20647

automagically commented on sdk/communication/communication-phone-numbers/CHANGELOG.md:
```diff
@@ -1,5 +1,11 @@
+
+### Features Added
+
+- Upgraded the package to use the latest version of Azure Core, bringing several improvements to the package's HTTP pipeline architecture.
```
Would be nice to note the version here for future reference

automagically commented on sdk/communication/communication-phone-numbers/src/phoneNumbersClient.ts:
```diff
@@ -130,9 +136,8 @@ export class PhoneNumbersClient {
-      const { _response, ...results } = await this.client.getByNumber(phoneNumber, updatedOptions);
-      return results;
-    } catch (e) {
+      return await this.client.phoneNumbers.getByNumber(phoneNumber, updatedOptions);
```
Correct if I'm wrong but by waiting this function call aren't you break the return type? or is this function returning a promise after you wait for it?

danielortega-msft replied:
I don't think that this changes the return type; even if the call is awaited, the return is still wrapped around a Promise. Also, the previous implementation also awaited the call; the only difference is that it assigned the return value to a variable first in order to filter out the `_response` property from the results (which is no longer necessary since that property is no longer used in Core V2).

automagically commented on sdk/communication/communication-phone-numbers/src/utils/customPipelinePolicies.ts:
```diff
@@ -0,0 +1,59 @@
+      const nextLink: string = response?.parsedBody?.nextLink;
+
+      if (nextLink && !nextLink.startsWith(host)) {
+        const absolutePath = host.endsWith("/") ? host + nextLink.substring(1) : host + nextLink;
```
Here you are doing two thing at once, fixing the `nextLink` var and creating the absolute path, for me it would make more sense to split this two steps, so it is easier to identify the logic to create the final path, and as a bonus you adhere more to the DRY principle.

danielortega-msft replied:
Do you mean something like this (considering that `nextLink` is changed so it can be reassigned):
```suggestion
        nextLink =  host.endsWith("/") ? nextLink.substring(1) : nextLink;
        const absolutePath = host + nextLink;
```

jeremymeng commented on sdk/communication/communication-phone-numbers/src/phoneNumbersClient.ts:
```diff
@@ -110,9 +109,16 @@ export class PhoneNumbersClient {
       },
     };
 
+    this.client = new PhoneNumbersGeneratedClient(url, {
```
A couple lines above, the code to handle `userAgentOptions` should be removed. https://github.com/Azure/autorest.typescript/wiki/%60core-http%60-dependency-migration-to-%60core-client%60-%60core-rest-pipeline%60#remove-the-code-to-handle-useragentoptions

danielortega-msft replied:
Thanks, I wasn't aware of that guide. I will update accordingly!

lucasrsant commented on sdk/communication/communication-phone-numbers/src/phoneNumbersClient.ts:
```diff
@@ -110,9 +109,16 @@ export class PhoneNumbersClient {
+
+    // These policies are temporary workarounds to address compatibility issues with Azure Core V2.
+    this.client.pipeline.addPolicy(phoneNumbersLroPolicy);
+    this.client.pipeline.addPolicy(createPhoneNumbersPagingPolicy(url));
```
Since the `authPolicy` is being stored in a variable before being passed to the `addPolicy` method, I'd recommend you to keep the same pattern and have the `createPhoneNumbersPagingPolicy` result stored in a variable as well, even though this is a temporary workaround.

lucasrsant commented on sdk/communication/communication-phone-numbers/src/utils/customPipelinePolicies.ts:
```diff
@@ -0,0 +1,59 @@
+      const nextLink: string = response?.parsedBody?.nextLink;
+
+      if (nextLink && !nextLink.startsWith(host)) {
+        const absolutePath = host.endsWith("/") ? host + nextLink.substring(1) : host + nextLink;
```
I'm not sure if the SDK is based in ES6, but if so, I'd recommend using string templates here.

lucasrsant commented on sdk/communication/communication-phone-numbers/test/internal/customPipelinePolicies.spec.ts:
```diff
@@ -0,0 +1,117 @@
+// Copyright (c) Microsoft Corporation.
```
I see you testing the policies individually here, but I'm wondering if we should test them in the client itself.

danielortega-msft commented:
@lucasrsant the phone number that is present in the recordings is actually the sanitized value. We even explicitly use it as a fake number in one of our tests: https://github.com/danielortega-msft/azure-sdk-for-js/blob/0b5233eec003cf06439612349d9023bf50fee51a/sdk/communication/communication-phone-numbers/test/public/get.spec.ts#L37.

You can check in the recorded client configuration that the actual phone numbers are replaced with it.
``` ts
const replaceableVariables: { [k: string]: string } = {
  COMMUNICATION_LIVETEST_STATIC_CONNECTION_STRING: "endpoint=https://endpoint/;accesskey=banana",
  INCLUDE_PHONENUMBER_LIVE_TESTS: "false",
  SKIP_UPDATE_CAPABILITIES_LIVE_TESTS: "false",
  COMMUNICATION_ENDPOINT: "https://endpoint/",
  AZURE_CLIENT_ID: "SomeClientId",
  AZURE_CLIENT_SECRET: "azure_client_secret",
  AZURE_TENANT_ID: "SomeTenantId",
  AZURE_PHONE_NUMBER: "+14155550100",
  COMMUNICATION_SKIP_INT_PHONENUMBERS_TESTS: "false",
};
```

danielortega-msft commented on sdk/communication/communication-phone-numbers/recordings/browsers/phonenumbersclient__get_phone_number/recording_can_get_a_purchased_phone_number.json:
```diff
@@ -12,18 +12,19 @@
+    "ms-cv": "owzk+pDQjEuVnbBCe6Qcnw.0",
     "request-context": "appId=",
-    "x-azure-ref": "0e73dYQAAAAD6/FN7jN4gT45wQANyU0NsU0FPMDNFREdFMDkwOQA5ZmM3YjUxOS1hOGNjLTRmODktOTM1ZS1jOTE0OGFlMDllODE=",
+    "strict-transport-security": "max-age=2592000",
```
Not sure if it's because of that reason, but there were failures using the prior recordings. Not all of them had to be recorded again, but I decided to run the whole test suite in record mode just to make sure everything works as expected.

danielortega-msft commented:
@jeremymeng @automagically @lucasrsant I updated the PR based on your comments, can you please do another review?

deyaaeldeen commented:
> FYI, the following issue was created to track the LRO incompatibility: https://github.com/Azure/azure-sdk-for-js/issues/20647

The LRO bug has been fixed and I will make a release on Monday, 3/7.

jeremymeng commented on sdk/communication/communication-phone-numbers/test/internal/headers.spec.ts:
```diff
@@ -2,23 +2,24 @@
     httpClient: getPhoneNumberHttpClient,
-  });
-  let request: WebResourceLike;
+  } as PhoneNumbersClientOptions);
```
Curious why a cast is needed here?  `PhoneNumbersClientOptions` should extends from `core-client`'s `CommonClientOptions`.

azure-sdk commented:
API changes have been detected in `@azure/communication-phone-numbers`. You can review API changes [here](https://apiview.dev/Assemblies/Review/8c9be04c627e485baab46d8d932c3de3?diffOnly=True&diffRevisionId=66e03504803341acac4b37449cb9accc)

**API changes**
```diff
- export interface PhoneNumbersClientOptions extends PipelineOptions { }
+ export interface PhoneNumbersClientOptions extends CommonClientOptions { }
```

jeremymeng commented on sdk/communication/communication-phone-numbers/swagger/README.md:
```diff
@@ -17,12 +17,13 @@ model-date-time-as-string: false
 skip-enum-validation: true
 title: Phone Numbers Client
 v3: true
+use-core-v2: true
```
nit: `use-core-v2` is true now by default so this is not needed.

jeremymeng commented on sdk/communication/communication-phone-numbers/src/generated/src/phoneNumbersClientContext.ts:
```diff
@@ -32,48 +28,25 @@ export class PhoneNumbersClientContext extends coreHttp.ServiceClient {
+      requestContentType: "application/json; charset=utf-8"
+    };
+
+    const packageDetails = `azsdk-js-azure-communication-phone-numbers/1.2.0-beta.2`;
```
The `azure` after `azsdk-js-` is probably from the package-name in swagger/README.md.  Could you please use `@azure/communication-phone-numbers` instead?

jeremymeng reviewed (approved):
Looks good to me! I only have some minor comments.

danielortega-msft commented:
> Looks good to me! I only have some minor comments.

Thanks for the feedback @jeremymeng. I will merge this PR for now and, once the LRO fix is released, I will create a new one addressing your comments.
